### PR TITLE
feat: support custom names for generated k8s resources

### PR DESCRIPTION
**Pull request by arkodg**
Relates to https://github.com/envoyproxy/gateway/pull/3527

### CODE REVIEW

**Review by cnvergence — CHANGES_REQUESTED** [review]
probably we would need to also change it in kubernetes provider
https://github.com/envoyproxy/gateway/blob/607d8bcd9a833841eee057b263c0bdc06de0fbea/internal/provider/kubernetes/helpers.go#L151

**Comment by arkodg** [comment]
hey @cnvergence can you elaborate why thats needed, the above diff is changing the value right before setting the resource name ?

**Comment by arkodg** [comment]
oh @cnvergence I see the issue now, we are fetching based on name and not labels 😢 
https://github.com/envoyproxy/gateway/blob/cdd4e065d4a2a5e3657052664073972f06ac77dd/internal/provider/kubernetes/predicates.go#L467
&
https://github.com/envoyproxy/gateway/blob/cdd4e065d4a2a5e3657052664073972f06ac77dd/internal/provider/kubernetes/predicates.go#L483

cc @dprotaso this is the issue seen in https://github.com/knative-extensions/net-gateway-api/pull/738

File: internal/gatewayapi/helpers.go [hunk 1]
```diff
@@ -228,6 +228,15 @@ func GatewayClassOwnerLabel(name string) map[string]string {
 	return map[string]string{OwningGatewayClassLabel: name}
 }
 
+// OwnerLabels returns the owner labels based on the mergeGateways setting
+func OwnerLabels(gateway *gwapiv1.Gateway, mergeGateways bool) map[string]string {
```
Comment (cnvergence) [inline internal/gatewayapi/helpers.go:232]: 🚀

**Review by cnvergence — APPROVED** [review]
LGTM

**Comment by davem-git** [comment]
If this is addressed here, will it also fix when I rename the `envoyDeployment`:, the HPA `envoyHpa`? It no longer works it keeps the original generated name. There doesn't seem to be a way to patch that either

**Comment by arkodg** [comment]
> If this is addressed here, will it also fix when I rename the `envoyDeployment`:, the HPA `envoyHpa`? It no longer works it keeps the original generated name. There doesn't seem to be a way to patch that either

thanks for catching this ! ive fixed the deployment targetRef naming in hpa.
I havent changed the name on the hpa resource itself, that needs an API change, which can be handled in a future PR

**Review by zhaohuabing — COMMENTED** [review]
🚀

**Review by zhaohuabing — COMMENTED** [review]
🚀

**Review by zhaohuabing — APPROVED** [review]

**Comment by davem-git** [comment]
On this same issue. I've renamed the deployment by patching the name, this causes the gateway to break. While the deployment with 3 replicas I get this error 

```sh
  Last Transition Time:  2024-06-14T15:02:58Z
    Message:               The Gateway has been scheduled by Envoy Gateway
    Observed Generation:   1
    Reason:                Accepted
    Status:                True
    Type:                  Accepted
    Last Transition Time:  2024-06-14T15:03:07Z
    Message:               Deployment replicas unavailable
    Observed Generation:   1
    Reason:                NoResources
    Status:                False
    Type:                  Programmed
```

the gateway status is this 

```sh
NAME            CLASS           ADDRESS         PROGRAMMED   AGE
acme-envoy   acme-envoy   IPADDRESS   False        4m19s
```

**Comment by arkodg** [comment]
@davem-git can you raise a separate GH issue to highlight the issue you are facing

**Pull request merged**